### ISSUE TITLE
Still shows Add to cart even if salable quantity is 0

### ISSUE THREAD

**Issue by chequille** [magento/magento2#33819]
### Preconditions (*)

1. Magento 2.4.3
2.

### Steps to reproduce (*)

1. Set "Display out of stock products" to Yes
2. Place order with a product so salable qty is going to 0

### Expected result (*)

1. After placing order, add to cart should not be displayed anymore for product qith salable quantity 0
2.

### Actual result (*)

1. Add to cart button is still shown
2.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by chequille** [comment]
@magento give me 2.4-develop instance

**Comment by chequille** [comment]
Unbelievable, products where salable qty is 0 are still showing ADD TO CART button.
I reproduced this in the Magento instance.
Release notes of 2.4.3 says this is fixed, but it looks like it is not fixed.

What are you going to do on this?

**Comment by engcom-Echo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Echo** [comment]
Hi @chequille 
We tried to reproduce this issue on 2.4-develop & was unable to do so.
Here are the Steps we followed,
1. **Stores** > **Configuration** > **Catalog** > **Inventory** > **Stock Options**. Display Out of Stock Products: Choose **Yes** to enable displaying Magento out of stock products in the frontend
2. Create / Edit a product such that its quantity is 1 & place an order for the product from frontend such that saleable quantity will become 0.
<img width="1264" alt="Screenshot 2021-08-17 at 22 07 15" src="https://user-images.githubusercontent.com/51681435/129766504-ac055627-b834-4229-abc9-657c552a6e6b.png">
<img width="1363" alt="Screenshot 2021-08-17 at 22 08 29" src="https://user-images.githubusercontent.com/51681435/129766564-929e54bc-01d3-444a-809c-b6cf6b4431e7.png">
3. Once the order was placed , the product was displayed OUT OF STOCK on Search & Product Page.

<img width="1363" alt="Screenshot 2021-08-17 at 22 08 52" src="https://user-images.githubusercontent.com/51681435/129766907-24e78ee9-aac7-40bf-a291-0e877bbc3649.png">

So kindly recheck this issue on Magento 2.4-develop branch and if the issue is still reproducible, please provide more details or additional/missing steps regarding the same.
Thank you.

**Comment by chequille** [comment]
Made another try with "Impulse duffle"
![grafik](https://user-images.githubusercontent.com/15807794/129772632-db41bb1a-6dc0-4449-a1c8-7b4b81fba649.png)
Bought 100 pieces for product going to salable 0

![grafik](https://user-images.githubusercontent.com/15807794/129772862-47f076de-0c76-45d9-80d4-50c860c3bf78.png)
Going to the product page after order was placed.
And Add to cart button is there as ou see:
![grafik](https://user-images.githubusercontent.com/15807794/129773132-30f89a49-becc-4992-b701-a34932b62a11.png)
Sorry, but I am not stupid and I did the same as you did. Do not understand why Add to Cart button is not there anymore on your try.

**Comment by chequille** [comment]
Hi @engcom-Echo,

now I found the problem. As I mentioned before, you changed the inventory from 100 to 1 which I did not. I bought the 100 piece. I tried now to first edit the prodcut and save it and than making an order for 100 pieces. When I do so, add to cart button is nt shown anymore.

But this means, that the saving of a product does change something somewhere in the database.
Of course, I did not do this on my site as I use a copy of my 2.3.x database (of course changing url for the site) and do a magento setup:upgrade which works so far as it should.
If this is really the case as i described it I have to save all products at least once.

I will try this now and will report my results.

Regards,
Jürgen

**Comment by chequille** [comment]
Hi @engcom-Echo
unfortunately, on my site even the editing and saving a product again does not change something. On my site add to cart is shown.

but something must be wrong in any case, Try not to change the product and place an order with the quantity which is in the database of a fresh magento instance (I think they are all set to 100) and you will see, that Add to cart is shown.

What else could it be? As I said already in Magento 2.3.x everything worked fine. Therefore, I do not think that some of my extensions causes this behaviour.

Does somebody have any idea?
Regards
Jürgen

**Comment by engcom-Echo** [comment]
@magento give me 2.4-develop instance

**Comment by kemexyz** [comment]
I habe a magento 2.4.2 now updated to Magento 2.4.3 affected by the same issue.

Got a simple product with 
qty 1 
salable qty 0
reservation for qty 1 form an order

Add to cart button still showing after reindex, cache. 

All salabe check functions report the product is salable and in stock which is not correct.

**Comment by chequille** [comment]
Oh, somebody else also has this problem. In some point of view I am very happy about this :-) I am not the only one.

In the meantime I disabled all third party modules which I am using in my setup, but still wrong!!!!

**Comment by kemexyz** [comment]
@chequille Did you do a data migration from m1 ?

This issue does not happen with all of the products for me.

**Comment by chequille** [comment]
@kemexyz 
No, I did not data migration.
I am using database dump from my working 2.3.x installation and change of course the urls in core_config_data table and run magento setup:upgrade

This works pretty good, but maybe not regarding this issue. 
What I tried is to make a new product and did have a look with this new product. On my site, it still have the same behaviour.

So, no idea whats happening here.

In the meantime, I made my own block class which is using GetSalableQuantityDataBySku $getSalableQuantityDataBySku class to check if salable quantity is 0 or not.
The only disadvantage doing this is, that I have to make my own block not cachable, because otherwise my function to get the value for salable qty is not the correct one.
I do not see a real performance change, so if the guys here at magento will not find a solution, I will use it.
Unfortunately, there is another problem with swatches of configurable products, which are not crossed if salable qty is 0. There is as well another ticket from my site regarding this.
DId not find a solution for that yet.
Regards
Jürgen

**Comment by chequille** [comment]
@kemexyz 
By the way, I remember now, that I did a data migration when I switch from M1 to M2 last year in June.
Do you think this can be a problem?

**Comment by kemexyz** [comment]
@chequille I was thinking about it after testing a product that did not have the issue. But on a second test it was salable again even when salable qty was 0.

Just seems strange that we would be the only 2 people seeing this. I'm gonna try to test this on a clean M2.4.3 tonight.

**Comment by chequille** [comment]
@engcom-Echo
Cannot belileve that you set this issue to "cannot reproduce"
DID YOU READ WHAT I WROTE?

Once again: change Display Out of Stock Products to Yes and than place an order for a bag with qty 100 and have a look at this product. It is showing the Add to cart button as I said.
Do not change anything else. Do not edit the product and change qty, just place the order with qty 100 and you will see the ADD TO CART BUTTON.

Maybe one hint why it is not working on my site: I found, that the database view inventory_stock_1 shows my products as salable even if in the backend under products salable qty is already 0
If I  make the shipping for the orders and qty is going to 0, than inventory_stock_1 is going as well to 0

This seems the problem on my site and the question is why?

Please handle this issue as a serious one.

BR
Chequille

**Comment by engcom-Echo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Echo** [comment]
Hi @chequille,
First of all thank you for providing additional details for reproducing the issue, though we were unable to reproduce the issue at first instance with the steps you had provided, however with the new details we are able to reproduce the issue.
But even with the new steps the issues doesnt seems to reiterate every item, I was able to reproduce the issue only for the first time on a clean magento instance, for every other trial the feature seemed to be working fine. Please refer the video for the steps taken to reproduce the issue.

https://user-images.githubusercontent.com/51681435/130015373-fe5f41c4-cc98-4d8f-be90-38ef5344b379.mp4

**Comment by chequille** [comment]
Hi @engcom-Echo

I saw your video but on my 2.4.3 installation it does not behave the same. ADD to Cart button only dissapears if I made the shippment. So here is my video and I guess, that something went wrong with updating the inventory_stock_1 view or the tables for this view.
Video is a google drive link: https://drive.google.com/file/d/1WuwoxL4LD8rHjYyIK6wbPODntsE0cSSn/view?usp=sharing

**Comment by chequille** [comment]
Hi Guys,
I do not know if this helps to find why on my site inventory stock still have the value 1 for the status, but I provide the traces and sql queries. Maybe somebody sees what is going wrong.
The traces are from an order for roduct with ID 84 and 17 ieces were available before the order.

## 2021-08-20 10:34:01
## 2807 ## QUERY
SQL: INSERT  INTO `cataloginventory_stock_status` (`product_id`,`website_id`,`stock_id`,`qty`,`stock_status`) VALUES (?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE `qty` = VALUES(`qty`), `stock_status` = VALUES(`stock_status`)
BIND: array (
  0 => 84,
  1 => 0,
  2 => 1,
  3 => 17.0,
  4 => 1,
)
AFF: 1
TIME: 0.0024
TRACE: #1 Magento\Framework\DB\Logger\File[Magento\Framework\DB\Logger\LoggerAbstract]#0000000077619d6900000000148a1009#->getStats('query', 'INSERT  INTO `ca...', array(84, 0, 1, 17, 1), &Magento\Framework\DB\Statement\Pdo\Mysql#000000007761929d00000000148a1009#) called at [vendor/magento/framework/DB/Logger/File.php:67]
#2 Magento\Framework\DB\Logger\File#0000000077619d6900000000148a1009#->logStats('query', 'INSERT  INTO `ca...', array(84, 0, 1, 17, 1), &Magento\Framework\DB\Statement\Pdo\Mysql#000000007761929d00000000148a1009#) called at [vendor/magento/framework/DB/Logger/LoggerProxy.php:152]
#3 Magento\Framework\DB\Logger\LoggerProxy#0000000077619d0a00000000148a1009#->logStats('query', 'INSERT  INTO `ca...', array(84, 0, 1, 17, 1), &Magento\Framework\DB\Statement\Pdo\Mysql#000000007761929d00000000148a1009#) called at [vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:561]
#4 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor[Magento\Framework\DB\Adapter\Pdo\Mysql]#0000000077619d7700000000148a1009#->_query('INSERT  INTO `ca...', array(84, 0, 1, 17, 1)) called at [vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:627]
#5 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor[Magento\Framework\DB\Adapter\Pdo\Mysql]#0000000077619d7700000000148a1009#->query('INSERT  INTO `ca...', array(84, 0, 1, 17, 1)) called at [generated/code/Magento/Framework/DB/Adapter/Pdo/Mysql/Interceptor.php:95]
#6 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor#0000000077619d7700000000148a1009#->query('INSERT  INTO `ca...', array(84, 0, 1, 17, 1)) called at [vendor/magento/framework/DB/Adapter/Pdo/Mysql.php:2031]
#7 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor[Magento\Framework\DB\Adapter\Pdo\Mysql]#0000000077619d7700000000148a1009#->insertOnDuplicate('cataloginventory...', array(array('product_id' => 84, 'website_id' => 0, 'stock_id' => 1, 'qty' => 17, 'stock_status' => 1)), array('qty', 'stock_status')) called at [generated/code/Magento/Framework/DB/Adapter/Pdo/Mysql/Interceptor.php:374]
#8 Magento\Framework\DB\Adapter\Pdo\Mysql\Interceptor#0000000077619d7700000000148a1009#->insertOnDuplicate('cataloginventory...', array(array('product_id' => 84, 'website_id' => 0, 'stock_id' => 1, 'qty' => 17, 'stock_status' => 1)), array('qty', 'stock_status')) called at [vendor/magento/module-catalog-inventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php:352]
#9 Magento\CatalogInventory\Model\ResourceModel\Indexer\Stock\DefaultStock#000000007761929b00000000148a1009#->_updateIndexTable(array(array('product_id' => 84, 'website_id' => 0, 'stock_id' => 1, 'qty' => 17, 'stock_status' => 1))) called at [vendor/magento/module-catalog-inventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php:318]
#10 Magento\CatalogInventory\Model\ResourceModel\Indexer\Stock\DefaultStock#000000007761929b00000000148a1009#->_updateIndex(array(84)) called at [vendor/magento/module-catalog-inventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php:131]
#11 Magento\CatalogInventory\Model\ResourceModel\Indexer\Stock\DefaultStock#000000007761929b00000000148a1009#->reindexEntity(array(84)) called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock/AbstractAction.php:276]
#12 Magento\CatalogInventory\Model\Indexer\Stock\Action\Rows[Magento\CatalogInventory\Model\Indexer\Stock\AbstractAction]#000000007761928500000000148a1009#->doReindex(array(84)) called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock/AbstractAction.php:246]
#13 Magento\CatalogInventory\Model\Indexer\Stock\Action\Rows[Magento\CatalogInventory\Model\Indexer\Stock\AbstractAction]#000000007761928500000000148a1009#->Magento\CatalogInventory\Model\Indexer\Stock\{closure}() called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock/CacheCleaner.php:86]
#14 Magento\CatalogInventory\Model\Indexer\Stock\CacheCleaner#000000007761929500000000148a1009#->clean(array(84), &Closure#000000007761928800000000148a1009#) called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock/AbstractAction.php:247]
#15 Magento\CatalogInventory\Model\Indexer\Stock\Action\Rows[Magento\CatalogInventory\Model\Indexer\Stock\AbstractAction]#000000007761928500000000148a1009#->_reindexRows(array(84)) called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock/Action/Rows.php:30]
#16 Magento\CatalogInventory\Model\Indexer\Stock\Action\Rows#000000007761928500000000148a1009#->execute(array(84)) called at [vendor/magento/module-catalog-inventory/Model/Indexer/Stock.php:85]
#17 Magento\CatalogInventory\Model\Indexer\Stock\Interceptor[Magento\CatalogInventory\Model\Indexer\Stock]#000000007761928600000000148a1009#->executeList(array(84)) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#18 Magento\CatalogInventory\Model\Indexer\Stock\Interceptor#000000007761928600000000148a1009#->___callParent('executeList', array(array(84))) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#19 Magento\CatalogInventory\Model\Indexer\Stock\Interceptor#000000007761928600000000148a1009#->Magento\Framework\Interception\{closure}(array(84)) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#20 Magento\CatalogInventory\Model\Indexer\Stock\Interceptor#000000007761928600000000148a1009#->___callPlugins('executeList', array(array(84)), array(array('cache_cleaner_af...'))) called at [generated/code/Magento/CatalogInventory/Model/Indexer/Stock/Interceptor.php:41]
#21 Magento\CatalogInventory\Model\Indexer\Stock\Interceptor#000000007761928600000000148a1009#->executeList(array(84)) called at [vendor/magento/module-indexer/Model/Indexer.php:533]
#22 Magento\Indexer\Model\Indexer#000000007761930800000000148a1009#->reindexList(array(84)) called at [vendor/magento/module-indexer/Model/Indexer/DependencyDecorator.php:282]
#23 Magento\Indexer\Model\Indexer\DependencyDecorator#00000000776192d000000000148a1009#->reindexList(array(84)) called at [vendor/magento/framework/Indexer/AbstractProcessor.php:69]
#24 Magento\CatalogInventory\Model\Indexer\Stock\Processor[Magento\Framework\Indexer\AbstractProcessor]#000000007761995300000000148a1009#->reindexList(array(84)) called at [vendor/magento/module-catalog-inventory/Observer/ReindexQuoteInventoryObserver.php:66]
#25 Magento\CatalogInventory\Observer\ReindexQuoteInventoryObserver#00000000776192db00000000148a1009#->execute(&Magento\Framework\Event\Observer#000000007761930e00000000148a1009#) called at [vendor/magento/framework/Event/Invoker/InvokerDefault.php:88]
#26 Magento\Framework\Event\Invoker\InvokerDefault#0000000077619c2400000000148a1009#->_callObserverMethod(&Magento\CatalogInventory\Observer\ReindexQuoteInventoryObserver#00000000776192db00000000148a1009#, &Magento\Framework\Event\Observer#000000007761930e00000000148a1009#) called at [vendor/magento/framework/Event/Invoker/InvokerDefault.php:74]
#27 Magento\Framework\Event\Invoker\InvokerDefault#0000000077619c2400000000148a1009#->dispatch(array('instance' => 'Magento\CatalogI...', 'name' => 'inventory'), &Magento\Framework\Event\Observer#000000007761930e00000000148a1009#) called at [vendor/magento/framework/Event/Manager.php:66]
#28 Magento\Framework\Event\Manager#0000000077619c2b00000000148a1009#->dispatch('sales_model_serv...', array('order' => &Magento\Sales\Model\Order\Interceptor#00000000776191b100000000148a1009#, 'quote' => &Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#)) called at [generated/code/Magento/Framework/Event/Manager/Proxy.php:95]
#29 Magento\Framework\Event\Manager\Proxy#0000000077619d2500000000148a1009#->dispatch('sales_model_serv...', array('order' => &Magento\Sales\Model\Order\Interceptor#00000000776191b100000000148a1009#, 'quote' => &Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#)) called at [vendor/magento/module-quote/Model/QuoteManagement.php:581]
#30 Magento\Quote\Model\QuoteManagement\Interceptor[Magento\Quote\Model\QuoteManagement]#00000000776198da00000000148a1009#->submitQuote(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#, array()) called at [vendor/magento/module-quote/Model/QuoteManagement.php:456]
#31 Magento\Quote\Model\QuoteManagement\Interceptor[Magento\Quote\Model\QuoteManagement]#00000000776198da00000000148a1009#->submit(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#32 Magento\Quote\Model\QuoteManagement\Interceptor#00000000776198da00000000148a1009#->___callParent('submit', array(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#)) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#33 Magento\Quote\Model\QuoteManagement\Interceptor#00000000776198da00000000148a1009#->Magento\Framework\Interception\{closure}(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#34 Magento\Quote\Model\QuoteManagement\Interceptor#00000000776198da00000000148a1009#->___callPlugins('submit', array(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#), array(array('validate_purchas...', 'update_bundle_qu...', 'coupon_uses_incr...'))) called at [generated/code/Magento/Quote/Model/QuoteManagement/Interceptor.php:68]
#35 Magento\Quote\Model\QuoteManagement\Interceptor#00000000776198da00000000148a1009#->submit(&Magento\Quote\Model\Quote\Interceptor#0000000077619a3900000000148a1009#) called at [vendor/magento/module-quote/Model/QuoteManagement.php:414]
#36 Magento\Quote\Model\QuoteManagement\Interceptor[Magento\Quote\Model\QuoteManagement]#00000000776198da00000000148a1009#->placeOrder(3084, NULL) called at [generated/code/Magento/Quote/Model/QuoteManagement/Interceptor.php:50]
#37 Magento\Quote\Model\QuoteManagement\Interceptor#00000000776198da00000000148a1009#->placeOrder(3084) called at [vendor/magento/module-checkout/Model/PaymentInformationManagement.php:125]
#38 Magento\Checkout\Model\PaymentInformationManagement\Interceptor[Magento\Checkout\Model\PaymentInformationManagement]#000000007761901100000000148a1009#->savePaymentInformationAndPlaceOrder(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#39 Magento\Checkout\Model\PaymentInformationManagement\Interceptor#000000007761901100000000148a1009#->___callParent('savePaymentInfor...', array(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#)) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#40 Magento\Checkout\Model\PaymentInformationManagement\Interceptor#000000007761901100000000148a1009#->Magento\Framework\Interception\{closure}(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#41 Magento\Checkout\Model\PaymentInformationManagement\Interceptor#000000007761901100000000148a1009#->___callPlugins('savePaymentInfor...', array(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#), array(array('validate-agreeme...'))) called at [generated/code/Magento/Checkout/Model/PaymentInformationManagement/Interceptor.php:23]
#42 Magento\Checkout\Model\PaymentInformationManagement\Interceptor#000000007761901100000000148a1009#->savePaymentInformationAndPlaceOrder(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#)
#43 call_user_func_array(array(&Magento\Checkout\Model\PaymentInformationManagement\Interceptor#000000007761901100000000148a1009#, 'savePaymentInfor...'), array(3084, &Magento\Quote\Model\Quote\Payment#000000007761903200000000148a1009#, &Magento\Quote\Model\Quote\Address\Interceptor#000000007761900600000000148a1009#)) called at [vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php:95]
#44 Magento\Webapi\Controller\Rest\SynchronousRequestProcessor#0000000077619e5800000000148a1009#->process(&Magento\Framework\Webapi\Rest\Request\Proxy#0000000077619d4d00000000148a1009#) called at [vendor/magento/module-webapi/Controller/Rest.php:188]
#45 Magento\Webapi\Controller\Rest\Interceptor[Magento\Webapi\Controller\Rest]#0000000077619ed700000000148a1009#->dispatch(&Magento\Framework\App\Request\Http#0000000077619caa00000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:58]
#46 Magento\Webapi\Controller\Rest\Interceptor#0000000077619ed700000000148a1009#->___callParent('dispatch', array(&Magento\Framework\App\Request\Http#0000000077619caa00000000148a1009#)) called at [vendor/magento/framework/Interception/Interceptor.php:138]
#47 Magento\Webapi\Controller\Rest\Interceptor#0000000077619ed700000000148a1009#->Magento\Framework\Interception\{closure}(&Magento\Framework\App\Request\Http#0000000077619caa00000000148a1009#) called at [vendor/magento/framework/Interception/Interceptor.php:153]
#48 Magento\Webapi\Controller\Rest\Interceptor#0000000077619ed700000000148a1009#->___callPlugins('dispatch', array(&Magento\Framework\App\Request\Http#0000000077619caa00000000148a1009#), array(array('append_no_store_...'), array('webapiContorller...', 'configHash'))) called at [generated/code/Magento/Webapi/Controller/Rest/Interceptor.php:23]
#49 Magento\Webapi\Controller\Rest\Interceptor#0000000077619ed700000000148a1009#->dispatch(&Magento\Framework\App\Request\Http#0000000077619caa00000000148a1009#) called at [vendor/magento/framework/App/Http.php:116]
#50 Magento\Framework\App\Http\Interceptor[Magento\Framework\App\Http]#0000000077619c9100000000148a1009#->launch() called at [generated/code/Magento/Framework/App/Http/Interceptor.php:23]
#51 Magento\Framework\App\Http\Interceptor#0000000077619c9100000000148a1009#->launch() called at [vendor/magento/framework/App/Bootstrap.php:264]
#52 Magento\Framework\App\Bootstrap#0000000077619c3f00000000148a1009#->run(&Magento\Framework\App\Http\Interceptor#0000000077619c9100000000148a1009#) called at [pub/index.php:29]

**Comment by chequille** [comment]
In fact, on my site in the template of the module Magento_Catalog  product\list.phtml the method $_product->isSaleable() returns true even if salable quantity is 0

Does somebody have an idea to find out why this is the case, meaning how to debug it? I found this method in module-catalog\model\Product.php and compared it with my working 2.3.5-p2 installation but could not find any difference.

So, could somebody explain which functions/methods are called here?
Where does MSI come in at all?

Any help would be very much appreciated. Do not want to wait again for the next release.

Thanks and best regards.
Chequille

**Comment by chequille** [comment]
@magento give me 2.4-develop instance

**Comment by IanFiretoys** [comment]
just wanted to chime in and say that we're experiencing this as well on 2.4.2 and 2.4.3 (we are using a database that was migrated from magento 1.9.3 in 2019 if that helps narrow it down).

we're having this issue in simples and configs where there's a saleable qty of 0 but a stock qty of >0 and a reservation for an order that hasn't been shipped yet. once the order gets shipped the logic of setting the simple product to out of stock seems to work. but it looks to us like it's **only** using the stock qty and not the saleable qty for this logic. 

it's causing the most issues for customers in configs where the actual quantity in stock is less evident. whereas simples tend to display "0 in stock" and add to cart.

**Comment by engcom-Alfa** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Alfa** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-November** [comment]
Hi @chequille , I verified the issue again on Magento 2.4-develop branch using same product "Impulse Duffle" by placing order with all quantity (say 100). Add to Cart button is removed only after re-indexing and clearing cache.
Without reindex - On refresing the front-end - Add to cart button is displayed.
After the indexers are reindexed and cache cleared - Add to Cart button is removed and "Out of Sock" status is displayed.
![image](https://user-images.githubusercontent.com/60198231/135076621-457fef75-7197-440b-815e-ddccf93ea2e7.png)

![image](https://user-images.githubusercontent.com/60198231/135076576-0cc767fc-43bd-4048-92b6-44dc8cdeb5f1.png)

**Comment by chequille** [comment]
As I said all the time !!!!!!
What are you going to do?
It is a bug!!!!!!

**Comment by jackjimmy** [comment]
need to disabled saleable quantity related modules. or you need to truncate reservation table from magento. as we all know saleable quantity change instantly while quantity change after shipment generate. magento2 give option inside inventory setting change quantity when order placed but it's not useful . so there is temporary solution is truncate reservation table from database or disabled all inventory modules.

**Comment by engcom-November** [comment]
Verified the issue again on Magento 2.4-develop branch by placing order with all quantity (say 100).
When Indexers are set to "Update by Schedule" (System - Index Management) and placed Order with all salable quantity - "Add to Cart" button is not getting removed in Product Details page even after Indexers are updated.
Only on on Catalog Search Results page - "Out of stock" status is displayed. Hence this issue should be fixed.
![image](https://user-images.githubusercontent.com/60198231/135274128-d2dfb167-0cc1-4615-8737-63e176c3b38e.png)
![image](https://user-images.githubusercontent.com/60198231/135274183-72c3cca2-1563-49f0-8852-e04bad71c613.png)
![image](https://user-images.githubusercontent.com/60198231/135274255-09b7fc7c-58d9-46b2-be0a-0f1567dbdbd9.png)
![image](https://user-images.githubusercontent.com/60198231/135274327-caed5dd4-0cb8-408b-80cb-ba01a031616b.png)

**Comment by chequille** [comment]
Main Problem is, that metho isSaleable is not returning false if salable quantity is 0
A part of the solution would be Pull request issue #33799
This is change in Magento/InventoryIndexer/Model/IsProductSalable.php
I am saying part of solution because it is only working if the view page of the product or the view page of the category is not cached.

On my site I set one block in calatog_product_view.xml and in catalog_category_view.xml to cahcable="false"

WIth this, and the change in the above mentioned class "add to cart button" is not shown anymore if salable quantity is 0

If doing so no reindexing or cleaning cache is necessary. But the page is not cached and therfore a little bit slower.

Therefore, what must be found is to clean cache if salable quantity is going to 0 after an order was placed.

**Comment by kemexyz** [comment]
Is this not related or solved by #33911 ? Can't test it because the solution is not publicly available for #33911

**Comment by tgreeve** [comment]
@engcom-Echo @engcom-November @chequille 

I found a reliable way to reproduce this issue in a fresh installed magento version 2.4.3-p1.

There should be a magento process listening to queue "inventory.reservations.updateSalabilityStatus" to mark the product out-of-stock in storefront (storefront-out-of-stock). When the process is running, the product will be marked as out-of-stock automatically when salable quantity reached zero. The process is only created by the cronjob command (./bin/magento cron:run) when "Display Out of Stock Products" is set to "yes". 

You can find the process by `top` or `ps` command, the command is 

`/usr/local/bin/php /var/www/html/bin/magento queue:consumers:start inventory.reservations.updateSalabilityStatus --single-thread --max-messages=10000`

Killing the process or the process isn't running, the product will not be marked as out-of-stock. 

Hope this help.

**Comment by tgreeve** [comment]
After code review, storefront product Out of Stock status depends on 2 database columns when using Default Stock.

1.  cataloginventory_stock_item => is_in_stock
2.  cataloginventory_stock_status => stock_status

Number 2 defines the storefront behavior, so whenever this value is set to 0, the storefront will show the product as out of stock.
Number 1 is the DB representation of the product "Stock Status" everyone see on the CMS product configuration page.

So consider the following situation:

When the last item of the product is sold, MQ consumer process the message in queue "inventory.reservations.updateSalabilityStatus" and set Number 2 to 0, thus storefront of the product will show Out of Stock. 

This should be expected scenario in the development team.

But, when Stock indexer is triggered and Number 1 should be 1, Stock indexer will sync the 1 value in Number 1 to Number 2, thus the storefront Out of Stock status will be removed and the product page will show Add to cart icon.

The source code where Stock indexer determines the stock status can be found [here](https://github.com/magento/magento2/blob/2.4.3-release/app/code/Magento/CatalogInventory/Model/ResourceModel/Indexer/Stock/DefaultStock.php#L379-L393). 

The Stock indexer can be triggered by Schedule or by a Save action. When the indexer is set to be triggered on Save, there are numerous possibilities that it will roll back the Out of Stock status done by the queue consumer, including a CMS user modifying the product, an extension observer modifying custom attributes......

So basically, the Default Stock indexer ruined the MSI Out of Stock mechanism in a real production Magento setup.

**Comment by gjportegies** [comment]
@chequille we're experiencing exactly the same issue on a Magento 2.4.3 instance which is originally migrated from a Magento 1.9.4.3 instance.

**Comment by nkajic** [comment]
Hi @tgreeve thank you for your time describing problem.

I have same problem and see at least 5+ duplicate tickets closed without solution.
How did you manage to solve this? Do you have some patch you can share with rest of us?

Thanks, N

**Comment by chequille** [comment]
Hi @nkajic 
see my comment above from https://github.com/magento/magento2/issues/33819#issuecomment-930529339
This helps on my site to get it almost working except of the fact, that I have to make catalog and product page not cachable.

**Comment by nkajic** [comment]
Hi @chequille 

I have managed to get it working on 2.3.4-p1 by properly configuring consumers. #35105

But on earlier version there was definetly some other problem involved. Upgrading to 2.4.3-p1 and configuring consumers solved my problem.

I hope it hepls.

Thanks,
N

**Comment by chequille** [comment]
Hi @nkajic,
first, thanks for this info.
Ok, I will try this on my local Ubuntu VM.

But if this is the only solution, somebody who does not have ssh acces to the server will never get it working, right?
So far , I am in lucky, as my production site is on a VServer and I have ssh access.

As well I have a question regaring the consumers. You mentioned "properly configuring consumers", but what does it mean properly configured? Which consumers I have to run? Do I have to configure something else somewhere?

Do you have a description of everything you did? If yes, can you share it here?

Would be great.
Best regards,
Chequille

**Comment by dphilipps** [comment]
Hi @tgreeve 
I agree with your assessment about the indexer ruining the stock status.  I raised this very issue in September 2021 https://github.com/magento/magento2/issues/34127 and it was eventually closed as a duplicate issue when it isn't - frustrating!  With properly configured consumers products are correctly listed as out of stock when the saleable quantity hits 0 (there was an issue previously preventing this working properly).  The separate issues as you mentioned is is the stock status is then reverted when a re-index is run (I would reproduce this by running a CLI reindex).

**Comment by shomiadarshlt** [comment]
@magento I am working on this.

**Comment by shomiadarshlt** [comment]
@magento give me 2.4-develop instance

**Comment by shomiadarshlt** [comment]
Hi @chequille ,
 
I tried replicating the issue in magento 2.4 develop instance but could not replicate .
The steps followed were :

- Set "Display out of stock products" to Yes in Stores->Configuration->Catalog->Inventory
- Placed an  order with a product so  that salable quantity becomes 0 
![image](https://github.com/magento/magento2/assets/98819214/d584f69a-0757-4ba0-8800-c425b0307b0a)
- There is no "Add to cart" button displayed for the product 
![image](https://github.com/magento/magento2/assets/98819214/aebed099-8777-48fa-a167-5460d79af77a)

Are there any steps that have been missed in replicating the issue ?
Request you to help in this regard .

Thanks
Shomi

**Comment by JonathanHelvey** [comment]
This is still happening on our store with Magento2.4.4-p5 with simple products. See the images below. I have not reproduced this on a Vanilla Magento2 instance, only on my store. Steps to reproduce. 
1. Display out-of-stock items to - YES.
2. Manage Stock to YES. 
3. No backordes. 
4. Add All inventory of product to Cart.
5. Checkout with Free Shipping enabled. 
6. Reindex 

The Product now shows in admin Quantity of 1 Salable Quantity 0. With the Product/order being in processing and waiting to be shipped. However on the frontend the product is still showing the 'ADD TO CART' button and showing in stock in both frontend and admin.  

<img width="1815" alt="Screenshot 2023-08-29 at 6 30 03 PM" src="https://github.com/magento/magento2/assets/33816020/eb3e2d01-8f0f-43ac-9b5e-954704eaf56e">


<img width="1920" alt="Screenshot 2023-08-29 at 6 24 08 PM" src="https://github.com/magento/magento2/assets/33816020/03a3ef35-7c68-4cd4-a541-bfd8c3f9097f">

**Comment by engcom-Bravo** [comment]
Hi @chequille,

Thanks for your reporting and collaboration.

We have verified the issue in Latest 2.4-develop instance and the issue is no more reproducible.Kindly refer the screenshots.

<img width="1436" alt="Screenshot 2024-07-25 at 14 27 41" src="https://github.com/user-attachments/assets/8379c609-9527-4fe5-be5a-04b04f2c2695">



<img width="1436" alt="Screenshot 2024-07-25 at 14 28 01" src="https://github.com/user-attachments/assets/9f32d650-4d14-4e7d-ae5c-55b96955c368">


There is no Add to Cart button for the products having salable quantity 0.

Hence we are closing this issue.

Thanks.

**Comment by DevBigBoy** [comment]
I have this Issue: "Salable Quantity = 0" but the Product Still Shows as "In Stock"

![Image](https://github.com/user-attachments/assets/5864d000-f1b6-4e39-9c36-a02a389770f3)

how can I solve this
